### PR TITLE
update dreambooth lora to work with IF stage II

### PR DESCRIPTION
**Pull request by williamberman**

### CODE REVIEW

File: examples/dreambooth/train_dreambooth_lora.py [hunk 1]
```diff
@@ -1143,8 +1157,24 @@ def compute_text_embeddings(prompt):
...
                     )
 
+                if unet.config.in_channels > channels:
+                    needed_additional_channels = unet.config.in_channels - channels
+                    additional_latents = randn_tensor(
+                        (bsz, needed_additional_channels, height, width),
+                        device=noisy_model_input.device,
+                        dtype=noisy_model_input.dtype,
+                    )
+                    noisy_model_input = torch.cat([additional_latents, noisy_model_input], dim=1)
```
Comment (williamberman) [inline examples/dreambooth/train_dreambooth_lora.py:1167]: if stage II takes an additional 3 channels of noise concatenated with the noised image

File: examples/dreambooth/train_dreambooth_lora.py [hunk 2]
```diff
@@ -1248,9 +1278,18 @@ def compute_text_embeddings(prompt):
...
+                    images = [
+                        pipeline(**pipeline_args, generator=generator).images[0]
+                        for _ in range(args.num_validation_images)
+                    ]
+                else:
+                    images = []
+                    for validation_upsample_image in args.validation_images:
+                        image = Image.open(validation_upsample_image)
+                        image_upsampled = pipeline(**pipeline_args, image=image, generator=generator).images[0]
+                        images.append(image_upsampled)
```
Comment (williamberman) [inline examples/dreambooth/train_dreambooth_lora.py:1292]: Allow using input images for validation

File: examples/dreambooth/train_dreambooth_lora.py [hunk 3]
```diff
@@ -1143,8 +1157,24 @@ def compute_text_embeddings(prompt):
...
+                if args.class_labels_conditioning == "timesteps":
+                    class_labels = timesteps
+                else:
+                    class_labels = None
+
                 # Predict the noise residual
-                model_pred = unet(noisy_model_input, timesteps, encoder_hidden_states).sample
+                model_pred = unet(
+                    noisy_model_input, timesteps, encoder_hidden_states, class_labels=class_labels
+                ).sample
```
Comment (williamberman) [inline examples/dreambooth/train_dreambooth_lora.py:1177]: IF stage II uses timesteps as "class_labels" conditioning. Note that class_labels is overloaded and this just means embeddings added with image latents

File: examples/dreambooth/train_dreambooth_lora.py [hunk 4]
```diff
@@ -1248,9 +1278,18 @@ def compute_text_embeddings(prompt):
...
+                    images = [
+                        pipeline(**pipeline_args, generator=generator).images[0]
+                        for _ in range(args.num_validation_images)
+                    ]
+                else:
+                    images = []
+                    for image in args.validation_images:
+                        image = Image.open(image)
+                        image = pipeline(**pipeline_args, image=image, generator=generator).images[0]
+                        images.append(image)
```
Comment (williamberman) [inline examples/dreambooth/train_dreambooth_lora.py:1292]: Allow using input images for validation

**Review by patrickvonplaten — COMMENTED** [review]
Can we add some tests for the LoRA Loader Mixin plugin? That'd be super nice :-)

File: examples/dreambooth/train_dreambooth_lora.py (hunk 1 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora.py:1167]: 🧠

File: examples/dreambooth/train_dreambooth_lora.py (hunk 3 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora.py:1177]: Timesteps as class labels ...

That's confusing to me.

**Review by sayakpaul — APPROVED** [review]
LGTM!

**Comment by sayakpaul** [comment]
> Can we add some tests for the LoRA Loader Mixin plugin? That'd be super nice :-)

I think this can be tackled separately. Let's discuss here: https://github.com/huggingface/diffusers/pull/3550

File: examples/dreambooth/train_dreambooth_lora.py (hunk 3 above)
Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth_lora.py:1177]: Yeah I also think the name was badly chosen! Sorry think it was me who introduced it for another use case. The policy from now on is a bit that `class_labels` should always be integer inputs (either timesteps for super res noise, or actual embedding indexes - for e.g. sdx4 upscaler). So this is def ok here for me

**Comment by williamberman** [comment]
Was hoping to get the lora tests in before merging but looks like sayak has a better handle on it in https://github.com/huggingface/diffusers/pull/3550 so will leave to the follow up on there 👍

**Comment by xiaohu2015** [comment]
@williamberman Do you test the training code, I think it has some problems: the input should be two noisy images (one for diffusion, one for upscaled noisy image), but here you always set the first noisy image to be a random noise.

**Comment by sayakpaul** [comment]
> Was hoping to get the lora tests in before merging but looks like sayak has a better handle on it in #3550 so will leave to the follow up on there 👍

@williamberman feel free to add a test and we can see if there's still an issue.

**Comment by williamberman** [comment]
Hey @xiaohu2015 yes correct! thanks for the catch, fwiw it does still train but yes that should be fixed
